core/src/main/scala/chisel3/RawModule.scala

```diff
@@ -176,13 +177,17 @@ abstract class RawModule extends BaseModule {
   }
   private[chisel3] val stagedSecretCommands = collection.mutable.ArrayBuffer[Command]()
 
-  private[chisel3] def secretConnection(left: Data, right: Data)(implicit si: SourceInfo): Unit = {
-    val rhs = (left.probeInfo.nonEmpty, right.probeInfo.nonEmpty) match {
-      case (true, true)                                 => ProbeDefine(si, left.lref, Node(right))
-      case (true, false) if left.probeInfo.get.writable => ProbeDefine(si, left.lref, RWProbeExpr(Node(right)))
-      case (true, false)                                => ProbeDefine(si, left.lref, ProbeExpr(Node(right)))
-      case (false, true)                                => Connect(si, left.lref, ProbeRead(Node(right)))
-      case (false, false)                               => Connect(si, left.lref, Node(right))
+  private[chisel3] def secretConnection(left: BaseType, right: BaseType)(implicit si: SourceInfo): Unit = {
+    val rhs = (left, right) match {
+      case (lprop: Property[_], rprop: Property[_]) => PropAssign(si, lprop.lref, Node(rprop))
+      case (ldata: Data, rdata: Data) =>
+        (ldata.probeInfo.nonEmpty, rdata.probeInfo.nonEmpty) match {
+          case (true, true)                                  => ProbeDefine(si, ldata.lref, Node(rdata))
+          case (true, false) if ldata.probeInfo.get.writable => ProbeDefine(si, ldata.lref, RWProbeExpr(Node(rdata)))
+          case (true, false)                                 => ProbeDefine(si, ldata.lref, ProbeExpr(Node(rdata)))
+          case (false, true)                                 => Connect(si, ldata.lref, ProbeRead(Node(rdata)))
+          case (false, false)                                => Connect(si, ldata.lref, Node(rdata))
```

Comment:
should there be a non-exhaustive match case here?

Reply:
Do you mean like `case (_, _)` for the outer `match`? I didn't change the inner `match`, which I think is already exhaustive. If i'm understanding, yeah that makes sense, I can at least throw a meaningful exception in that case.

Reply:
I added an exception base case, but I think it might actually be impossible to hit--the Scala compiler should ensure you're either boring Data to Data, or Property to Property.

Reply:
I guess it would come up if we had a new kind of BaseType